substate-store-impls/src/substate_database_staging.rs

```diff
@@ -0,0 +1,282 @@
+                    }) => Box::new(
+                        new_substate_values
+                            .iter()
+                            .map(|(sort_key, substate_value)| {
+                                (sort_key.clone(), substate_value.clone())
+                            })
+                            .sorted(),
+                    ),
+                    // There are some changes that need to be overlayed.
+                    Some(PartitionDatabaseUpdates::Delta { substate_updates }) => {
+                        let underlying = self.root.list_entries_from(partition_key, from_sort_key);
+                        let overlaying = substate_updates
+                            .iter()
+                            .map(|(sort_key, database_update)| match database_update {
+                                DatabaseUpdate::Set(substate_value) => {
+                                    (sort_key.clone(), Some(substate_value.clone()))
+                                }
+                                DatabaseUpdate::Delete => (sort_key.clone(), None),
+                            })
+                            .sorted_by(|(sort_key_a, _), (sort_key_b, _)| {
+                                sort_key_a.cmp(sort_key_b)
+                            });
+                        Box::new(OverlayingIterator::new(underlying, overlaying))
+                    }
+                    // Overlay doesn't contain anything for the provided partition number. Return an
+                    // iterator over the data in the root store.
+                    None => self.root.list_entries_from(partition_key, from_sort_key),
+                }
+            }
+            // Overlay doesn't contain anything for the provided node key. Return an iterator over
+            // the data in the root store.
+            None => self.root.list_entries_from(partition_key, from_sort_key),
+        }
+    }
+}
+
+impl<'s, S> CommittableSubstateDatabase for SubstateDatabaseStaging<'s, S> {
+    fn commit(&mut self, database_updates: &DatabaseUpdates) {
+        merge_database_updates(&mut self.overlay, database_updates.clone())
+    }
+}
+
+impl<'s, S> ListableSubstateDatabase for SubstateDatabaseStaging<'s, S>
+where
+    S: ListableSubstateDatabase,
+{
+    fn list_partition_keys(&self) -> Box<dyn Iterator<Item = DbPartitionKey> + '_> {
+        Box::new(
+            self.root
+                .list_partition_keys()
+                .chain(self.overlay.node_updates.iter().flat_map(
+                    |(node_key, NodeDatabaseUpdates { partition_updates })| {
+                        partition_updates
+                            .keys()
+                            .map(|partition_num| DbPartitionKey {
+                                node_key: node_key.clone(),
+                                partition_num: *partition_num,
+                            })
+                    },
+                ))
+                .unique()
+                .sorted(),
```

Comment:
We can't use `unique()` and `sorted()` here - this sticks all keys into memory at once from the underlay store.

Instead, we should use an overlayed iterator, like inside of `list_entries_from`

Reply:
Sounds good, I've switched the impl to use `OverlayingIterator`.